File: backend-sdk/JavaScript/src/TokenGenerator.js
import CryptoJS from 'crypto-js';
import fs from 'fs/promises';

export class TokenGenerator {
    /**
     * Initialize token generator
     * 初始化令牌生成器
     * @param {string|object} config Configuration file path or configuration object / 配置文件路径或配置对象
     */
    constructor(config) {
        this.config = null;
        this.initialize(config);
    }

    /**
     * Initialize configuration
     * 初始化配置
     * @param {string|object} config Configuration file path or configuration object / 配置文件路径或配置对象
     */
    async initialize(config) {
        this.config = config;

        // Validate configuration / 验证配置
        this.validateConfig();
    }

    /**
     * Validate configuration
     * 验证配置
     */
    validateConfig() {
        const { crypto, server } = this.config;
        if (!crypto || !server) {
            throw new Error('Invalid config: missing crypto or server section');
        }
        if (crypto.method !== 'aes') {
            throw new Error('Only AES encryption is supported');
        }
        if (!crypto.aes_key || !crypto.aes_iv_seed) {
            throw new Error('Invalid crypto config: missing aes_key or aes_iv_seed');
        }
    }

    /**
     * Generate initialization vector
     * 生成初始化向量
     * @returns {WordArray} Generated IV / 生成的 IV
     */
    generateIV() {
        // Generate random IV / 生成随机 IV
        const iv = CryptoJS.lib.WordArray.random(16);
        // Mix with IV seed / 使用 IV 种子进行混合
        const ivSeed = CryptoJS.enc.Utf8.parse(this.config.crypto.aes_iv_seed);
        const words = iv.words.map((word, i) => word ^ ivSeed.words[i]);
        return CryptoJS.lib.WordArray.create(words, 16);
    }

    /**
     * Create token data
     * 创建令牌数据
     * @param {object} options Token options / 令牌选项
     * @returns {object} Token data / 令牌数据
     */
    createToken({
        apiKey,
        maxCalls = 100,
        expireDays = 1,
        provider = 'dashscope',
        extInfo = ''
    }) {
        const now = new Date();
        const expireTime = new Date(now.getTime() + expireDays * 24 * 60 * 60 * 1000);

        return {
            id: `token-${Date.now()}`,
            api_key: apiKey,
            max_calls: maxCalls,
            expire_time: expireTime.toISOString(),
            created_at: now.toISOString(),
            provider: provider,
            ext_info: extInfo
        };
    }

    /**
     * Encrypt token data
     * 加密令牌数据
     * @param {object} tokenData Token data / 令牌数据
     * @returns {string} Encrypted token string / 加密后的令牌字符串
     */
    encryptToken(tokenData) {
        // Serialize token data / 序列化令牌数据
        const jsonData = JSON.stringify(tokenData);
        
        // Generate IV / 生成 IV
        const iv = this.generateIV();
        
        // Decode AES key / 解码 AES 密钥
        const key = CryptoJS.enc.Hex.parse(this.config.crypto.aes_key);
        
        // Encrypt data / 加密数据
        const encrypted = CryptoJS.AES.encrypt(jsonData, key, {
            iv: iv,
            mode: CryptoJS.mode.CBC,
            padding: CryptoJS.pad.Pkcs7
        });
        
        // Combine IV and encrypted data / 组合 IV 和加密数据
        const combined = CryptoJS.lib.WordArray.create()
            .concat(iv)
            .concat(encrypted.ciphertext);
        
        // Base64 URL safe encoding / Base64 URL 安全编码
        return CryptoJS.enc.Base64url.stringify(combined);
    }

    /**
     * Get server URL
     * 获取服务器 URL
     * @param {string} path API path / API 路径
     * @returns {string} Complete server URL / 完整的服务器 URL
     */
    getServerUrl(path = '') {
        const { host, port, base_path } = this.config.server;
        const basePath = base_path.endsWith('/') ? base_path : `${base_path}/`;
        const cleanPath = path.startsWith('/') ? path.slice(1) : path;
        if (path === 'health') {
            return `${host}:${port}/health`;
        }
        return `${host}:${port}${basePath}${cleanPath}`;
    }
} 

File: backend-sdk/Java/src/main/java/com/github/relayapi/sdk/TokenGenerator.java
package com.github.relayapi.sdk;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.bouncycastle.crypto.engines.AESEngine;
import org.bouncycastle.crypto.modes.CBCBlockCipher;
import org.bouncycastle.crypto.paddings.PKCS7Padding;
import org.bouncycastle.crypto.paddings.PaddedBufferedBlockCipher;
import org.bouncycastle.crypto.params.KeyParameter;
import org.bouncycastle.crypto.params.ParametersWithIV;
import org.bouncycastle.util.encoders.Base64;
import org.bouncycastle.util.encoders.Hex;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.HashMap;
import java.util.Map;

public class TokenGenerator {
    private final ObjectMapper objectMapper = new ObjectMapper();
    private final Config config;
    private final String hash;

    public TokenGenerator(Config config) {
        this.config = config;
        validateConfig();
        this.hash = generateConfigHash();
    }

    private void validateConfig() {
        if (config.getCrypto() == null || config.getServer() == null) {
            throw new IllegalArgumentException("Invalid config: missing crypto or server section");
        }
        if (!"aes".equals(config.getCrypto().getMethod())) {
            throw new IllegalArgumentException("Only AES encryption is supported");
        }
        if (config.getCrypto().getAesKey() == null || config.getCrypto().getAesIvSeed() == null) {
            throw new IllegalArgumentException("Invalid crypto config: missing aes_key or aes_iv_seed");
        }
    }

    private String generateConfigHash() {
        try {
            String data = config.getCrypto().getMethod() +
                    config.getCrypto().getAesKey() +
                    config.getCrypto().getAesIvSeed();
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(data.getBytes(StandardCharsets.UTF_8));
            return Hex.toHexString(hash);
        } catch (Exception e) {
            throw new RuntimeException("Failed to generate config hash", e);
        }
    }

    private byte[] generateIV() {
        try {
            byte[] iv = new byte[16];
            SecureRandom random = new SecureRandom();
            random.nextBytes(iv);
            
            // Mix with IV seed
            byte[] ivSeed = config.getCrypto().getAesIvSeed().getBytes(StandardCharsets.UTF_8);
            for (int i = 0; i < iv.length && i < ivSeed.length; i++) {
                iv[i] ^= ivSeed[i];
            }
            return iv;
        } catch (Exception e) {
            throw new RuntimeException("Failed to generate IV", e);
        }
    }

    public Map<String, Object> createToken(TokenOptions options) {
        Instant now = Instant.now();
        Map<String, Object> tokenData = new HashMap<>();
        tokenData.put("id", "token-" + now.toEpochMilli());
        tokenData.put("api_key", options.getApiKey());
        tokenData.put("max_calls", options.getMaxCalls());
        tokenData.put("expire_time", now.plus(options.getExpireSeconds(), ChronoUnit.SECONDS).toString());
        tokenData.put("created_at", now.toString());
        tokenData.put("provider", options.getProvider());
        tokenData.put("ext_info", options.getExtInfo());
        return tokenData;
    }

    public String encryptToken(Map<String, Object> tokenData) {
        try {
            // Convert token data to JSON
            String jsonData = objectMapper.writeValueAsString(tokenData);
            
            // Generate IV
            byte[] iv = generateIV();
            
            // Setup AES cipher
            byte[] key = Hex.decode(config.getCrypto().getAesKey());
            CBCBlockCipher blockCipher = new CBCBlockCipher(new AESEngine());
            PaddedBufferedBlockCipher cipher = new PaddedBufferedBlockCipher(blockCipher, new PKCS7Padding());
            cipher.init(true, new ParametersWithIV(new KeyParameter(key), iv));
            
            // Encrypt data
            byte[] inputBytes = jsonData.getBytes(StandardCharsets.UTF_8);
            byte[] outputBytes = new byte[cipher.getOutputSize(inputBytes.length)];
            int length = cipher.processBytes(inputBytes, 0, inputBytes.length, outputBytes, 0);
            length += cipher.doFinal(outputBytes, length);
            
            // Combine IV and encrypted data
            byte[] combined = new byte[iv.length + length];
            System.arraycopy(iv, 0, combined, 0, iv.length);
            System.arraycopy(outputBytes, 0, combined, iv.length, length);
            
            // Base64URL encode
            return base64UrlEncode(combined);
        } catch (Exception e) {
            throw new RuntimeException("Failed to encrypt token", e);
        }
    }

    private String base64UrlEncode(byte[] data) {
        String base64 = Base64.toBase64String(data);
        return base64.replace('+', '-')
                .replace('/', '_')
                .replace("=", "");
    }

    public String getServerUrl(String path) {
        String basePath = config.getServer().getBasePath();
        if (!basePath.endsWith("/")) {
            basePath += "/";
        }
        String cleanPath = path.startsWith("/") ? path.substring(1) : path;
        
        if ("health".equals(path)) {
            return String.format("%s:%d/health",
                    config.getServer().getHost(),
                    config.getServer().getPort());
        }
        
        return String.format("%s:%d%s%s",
                config.getServer().getHost(),
                config.getServer().getPort(),
                basePath,
                cleanPath);
    }

    public String getHash() {
        return hash;
    }
} 